tinygrad/codegen/kernel.py

```diff
@@ -71,17 +71,21 @@ def __init__(self, *ast:LazyOp, opts:Optional[CompilerOptions]=None):
     self.ast = ast
     self.lazyops = flatten([op.lazyops for op in self.ast])
 
-    # there's only allowed to be one reduceop
-    reduceops = [x for x in self.lazyops if x.op in ReduceOps]
-    assert len(dedup(reduceops)) <= 1, "max one reduce op in an ast"
-    self.reduceop = reduceops[0] if reduceops else None
+    def get_rchildren(x:LazyOp): return ([x] if x.op in ReduceOps else []) + [r for s in x.src for r in get_rchildren(s)]
```

Comment:
```py
import time
from tinygrad.codegen.linearizer import Linearizer
from tinygrad.engine.schedule import create_schedule
from tinygrad.tensor import Tensor

st = time.perf_counter()
a = Tensor([1,2,3,4])
for _ in range(24): a = a + a
r = a.sum()
ast = create_schedule([r.lazydata])[-1].ast
lin = Linearizer(*ast)
assert time.perf_counter()-st < 1.0
```

Reply:
I futzed around with it and the traversal seems to just be too slow in general so instead of reordering them when the kernel get initialized, I them render in the ast_parse, because ast_parse is recursing through the ast anyways.

I did have to make some changes to `uops.py` so I'm making some more tests to make sure things still linearize properly